agenta-backend/agenta_backend/services/event_db_manager.py

```diff
@@ -134,31 +117,314 @@ async def create_trace_span(payload: CreateSpan) -> str:
+        app_id (str): The ID of the app
+        pagination (PaginationParam): The data of the pagination param
+        filters_param (GenerationFilterParams): The data of the generation filter params
+        sorters (SorterParams): The data of the sorters param
 
     Returns:
-        List[Span]: the list of spans for the given trace
+        List[Span]: the list of spans for the given user
     """
 
-    user = await db_manager.get_user(user_uid)
+    # Apply pagination and sorting
+    skip, limit = get_pagination_skip_limit(pagination)
+    sort_direction = (
+        pymongo.ASCENDING if sorters.created_at == "asc" else pymongo.DESCENDING
+    )
 
-    # Get trace
-    trace = await TraceDB.find_one(
-        TraceDB.id == ObjectId(trace_id), TraceDB.user.id == user.id, fetch_links=True
+    # Fetch spans without pagination and sorting applied
+    base_spans_db = SpanDB.find(SpanDB.trace.app_id == app_id)
+
+    # Count of spans in db
+    filters_query = {}
+    if filters_param.environment is not None:
+        filters_query["environment"] = filters_param.environment
+    # elif filters_param.variant is not None: TODO: improve in observability sdk
+    #     filters_query["base_id"] = filters_param.variant
+    spans_count = await base_spans_db.find(filters_query, fetch_links=True).count()
+
+    # Fetch spans with pagination and sorting applied
+    spans_db = base_spans_db.find(fetch_links=True, skip=skip, limit=limit).sort(
+        [(SpanDB.created_at, sort_direction)]
+    )
+
+    # Filter based on trace_id or not
+    if filters_param.trace_id is not None:
+        spans_db = await spans_db.find_many(
+            SpanDB.trace.id == ObjectId(filters_param.trace_id), fetch_links=True
+        ).to_list()
+    else:
+        spans_db = await spans_db.to_list()
+
+    # Convert beanie documents to pydantic models and filter based on the filter_params
+    spans = await spans_to_pydantic(spans_db)
+    filtered_generations = filter(
+        partial(filters.filter_document_by_filter_params, filters_param), spans
+    )
+    if filters_param.trace_id:
+        return list(filtered_generations)
+    return get_paginated_data(list(filtered_generations), spans_count, pagination)
+
+
+async def fetch_generation_span_detail(span_id: str) -> SpanDetail:
+    """Get a generation span detail.
+
+    Args:
+        span_id (str): The ID of a span
+
+    Returns:
+        SpanDetail: span detail pydantic model
+    """
+
+    span_db = await SpanDB.find_one(SpanDB.id == ObjectId(span_id), fetch_links=True)
+    app_variant_db = await db_manager.fetch_app_variant_by_id(span_db.trace.variant_id)
+
+    return SpanDetail(
+        **{
+            "id": str(span_db.id),
+            "created_at": span_db.created_at.isoformat(),
+            "variant": {
+                "variant_id": str(app_variant_db.id),
+                "variant_name": app_variant_db.variant_name,
+                "revision": app_variant_db.revision,
```

Comment:
This looks wrong, no? You are getting the revision number of the latest variant in production, and not the revision of the one that was invoked.

Say I invoke the llm app two weeks ago when the latest revision is 1. If I call fetch_generation_span today, when the latest revision is 5, I would see 5 here. 

The solution is to save this information when instrumenting this data. In fact, we should make sure not to have any inference at this step. 
This means, the variant_name and revision, should be saved with the span

Reply:
Ah, yes. You're right.